services/apps/src/apps/service/apps/jobs/sharing.clj

```diff
@@ -0,0 +1,170 @@
+   :load-failure "unable to load permissions for {{analysis-id}}: {{detail}}"
+   :not-allowed  "insufficient privileges for analysis ID {{analysis-id}}"})
+
+(defn- job-sharing-success
+  [job-id job level]
+  {:analysis_id   job-id
+   :analysis_name (get-job-name job-id job)
+   :permission    level
+   :success       true})
+
+(defn- job-sharing-failure
+  [job-id job level reason]
+  {:analysis_id   job-id
+   :analysis_name (get-job-name job-id job)
+   :permission    level
+   :success       false
+   :error         {:error_code ce/ERR_BAD_REQUEST
+                   :reason     reason}})
+
+(defn- job-unsharing-success
+  [job-id job]
+  {:analysis_id   job-id
+   :analysis_name (get-job-name job-id job)
+   :success       true})
+
+(defn- job-unsharing-failure
+  [job-id job reason]
+  {:analysis_id   job-id
+   :analysis_name (get-job-name job-id job)
+   :success       false
+   :error         {:error_code ce/ERR_BAD_REQUEST
+                   :reason     reason}})
+
+(defn- job-sharing-msg
+  ([reason-code job-id]
+   (job-sharing-msg reason-code job-id nil))
+  ([reason-code job-id detail]
+   (render (job-sharing-formats reason-code)
+           {:analysis-id job-id
+            :detail (or detail "unexpected error")})))
+
+(defn- load-analysis-permissions
+  [user analysis-id]
+  (try+
+   (iplant-groups/load-analysis-permissions user [analysis-id])
+   (catch ce/clj-http-error? {:keys [body]}
+     (throw+ {:type   ::permission-load-failure
+              :reason (:grouper_result_message (service/parse-json body))}))))
+
+(defn- has-analysis-permission
+  [user job-id required-level]
+  (-> (iplant-groups/load-analysis-permissions user [job-id])
+      (iplant-groups/has-permission-level required-level job-id)))
+
+(defn- share-app-for-job
+  [apps-client sharer sharee job-id {:keys [app-id]}]
+  (when-not (.hasAppPermission apps-client sharee app-id "read")
+    (let [response (.shareAppWithUser apps-client {} sharee app-id "read")]
+      (when-not (:success response)
+        (get-in response [:error :reason] "unable to share app")))))
+
+(defn- share-output-folder
+  [sharer sharee {:keys [result-folder-path]}]
+  (try+
+   (data-info/share-path sharer result-folder-path sharee "read")
+   nil
+   (catch ce/clj-http-error? {:keys [body]}
+     (str "unable to share result folder: " (:error_code (service/parse-json body))))))
+
+(defn- do-job-sharing-steps
+  [apps-client sharer sharee job-id job level]
+  (or (share-app-for-job apps-client sharer sharee job-id job)
+      (share-output-folder sharer sharee job)
+      (iplant-groups/share-analysis job-id sharee level)))
+
+(defn- share-accessible-job
+  [apps-client sharer sharee job-id job level]
+  (if-let [failure-reason (do-job-sharing-steps apps-client sharer sharee job-id job level)]
+    (job-sharing-failure job-id job level failure-reason)
+    (job-sharing-success job-id job level)))
+
+(defn- share-extant-job
```

Comment:
Having `share-job*` and `share-job*` would work, but I wanted a finer level of granularity here along with function names that provided a little bit of documentation.